File: experiments/02-rn-as-sdk/rn-sdk/ios/RnSdk/SdkBridge.m
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>

/**
 * Swift 로 쓴 네이티브 모듈을 RN 에 등록합니다.
 * RN 의 모듈 등록이 Objective-C 매크로라서 이 파일이 반드시 필요합니다.
 *
 * SDK 로 배포할 때 주의할 점이 하나 더 있습니다.
 * RCT_EXTERN_MODULE 이 만드는 등록 코드는 Objective-C 카테고리 형태라,
 * 정적 링크 과정에서 참조되지 않으면 링커가 떼어냅니다.
 * 소비 앱의 OTHER_LDFLAGS 에 -ObjC 가 필요한 이유입니다.
 * podspec 에서 그걸 강제합니다.
 */
@interface RCT_EXTERN_MODULE (RnSdkBridge, RCTEventEmitter)

RCT_EXTERN_METHOD(close)

RCT_EXTERN_METHOD(finishWithResult : (NSDictionary *)result)

RCT_EXTERN_METHOD(getHostInfo
                  : (RCTPromiseResolveBlock)resolve rejecter
                  : (RCTPromiseRejectBlock)reject)

@end

File: experiments/02-rn-as-sdk/rn-sdk/ios/RnSdk/SdkBridge.swift
import Foundation
// React 는 SDK 내부 구현에만 씁니다. 공개 API 는 UIKit/Foundation 타입만 노출합니다.
// 일반 import 로 두면 BUILD_LIBRARY_FOR_DISTRIBUTION 이 만드는 .swiftinterface 에
// `import React` 가 그대로 실려 나가고, React 모듈이 없는 소비 앱에서
// "no such module 'React'" 로 빌드가 깨집니다.
@_implementationOnly import React

/// Android 의 SdkBridgeModule 과 이름을 맞췄습니다(`RnSdkBridge`).
/// JS 쪽 js/native/SdkBridge.js 가 두 플랫폼에서 이 이름으로 찾습니다.
@objc(RnSdkBridge)
final class SdkBridge: RCTEventEmitter {

    private static let themeChangedEvent = "themeChanged"
    private var isObserving = false

    override static func requiresMainQueueSetup() -> Bool { false }

    override func supportedEvents() -> [String]! { [Self.themeChangedEvent] }

    override func startObserving() {
        isObserving = true
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleThemeChanged(_:)),
            name: SdkReactHost.themeChangedNotification,
            object: nil
        )
    }

    override func stopObserving() {
        isObserving = false
        NotificationCenter.default.removeObserver(
            self,
            name: SdkReactHost.themeChangedNotification,
            object: nil
        )
    }

    @objc private func handleThemeChanged(_ notification: Notification) {
        guard isObserving, let theme = notification.userInfo?["theme"] as? String else { return }
        sendEvent(withName: Self.themeChangedEvent, body: ["theme": theme])
    }

    @objc func close() {
        SdkScreenCoordinator.shared.close()
    }

    @objc func finishWithResult(_ result: NSDictionary) {
        SdkScreenCoordinator.shared.finish(
            with: RnSdk.CartResult(
                productId: result["productId"] as? String ?? "",
                action: result["action"] as? String ?? ""
            )
        )
    }

    @objc func getHostInfo(
        _ resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        let config = SdkReactHost.shared.hostConfig
        resolve([
            "appName": config.appName,
            "appVersion": config.appVersion,
            "platform": "ios",
        ])
    }
}
